feat: add project endpoints

jan-zabloudil opened the pull request:
This pull request introduces the projects. The functionality for 'project invitations' and 'project members' will be added in separate pr.

jan-zabloudil commented on service/model/project.go:
```diff
@@ -0,0 +1,91 @@
+	Delete(ctx context.Context, id uuid.UUID) error
+}
+
+type Project struct {
```
Project contains `ReleaseAlertConfig` and `SlackConfig`.

`ReleaseAlertConfig` replaces the original concept of release message templates and determinates which information to include into release alert. I also decided to use `ReleaseAlert` instead of `ReleaseMessage`.

`SlackConfig` has only one field for now - channel id.

jan-zabloudil commented on service/model/release_alert.go:
For now only `ReleaseAlertConfig` is only used as part of Project model.

But I decided to define `ReleaseAlert` already in order to better imagine how ReleaseAlertConfig will be used. And therefore I created separete file `/service/model/release_alert`.

`ReleaseAlert` holds the alert data. `ReleaseMessageBuilder` will be used for building the alert and will accept `ReleaseAlertConfig` as param.

jan-zabloudil commented on service/project.go:
Project service functions do not have any authorization logic yet because project members feature will be implemented in the separate PR.

jan-zabloudil commented on service/model/project_test.go:
I only added tests for service model functions.

Do you think it also make sense to mock repositories and write tests for service functions?

jan-zabloudil commented on service/model/project.go:
```diff
@@ -0,0 +1,91 @@
+	return nil
+}
+
+func (p *Project) Validate() error {
```
Validate function enforces all business rules and is called when creating or updating project object. 

I prefer to not use any validator package in service layer.

jan-zabloudil commented on repository/model/project.go:
```diff
@@ -0,0 +1,94 @@
+		UpdatedAt:          updatedAt.Time,
+	}
+
+	if err := p.Validate(); err != nil {
```
When converting db object into service object I also perform validation

jan-zabloudil commented on transport/model/project.go:
I prefer to have a separate struct for each operation, such as CreateProjectRequest, UpdateProjectRequest, and ProjectResponse.

In the transport layer, I only perform input validation to ensure that the user submits the correct payload. Any business rules are validated in the service layer

jan-zabloudil commented on service/model/release_alert.go:
`ReleaseAlertConfig` is a part of the Project struct, currently not used outside this context. 

But to illustrate the potential use of Config, I already have created a `ReleaseAlert` struct and a builder for it. And therefore I put all the related code into `/service/model/release_alert`. The ReleaseAlert struct is designed to store alert data. The builder, which takes a Config as a parameter, builds the alert.

jan-zabloudil commented on service/project.go:
There is no authorization logic in service functions for now because project members feature will be added in separate PR.

jan-zabloudil commented on service/model/project_test.go:
I created tests only for service model functions for now.

Would you also mock repositories and write tests for service functions?

bolshakoff commented on transport/utils/validate.go:
```diff
@@ -20,6 +21,16 @@ func init() {
 	translator, _ = uni.GetTranslator("en")
 	_ = entrans.RegisterDefaultTranslations(Validate, translator)
+
+	Validate.RegisterTagNameFunc(func(fld reflect.StructField) string {
```
A bit confused about this change; especially because the `init()` function is not used anywhere.. 🤔

jan-zabloudil replied:
Should be called automatically when i import the package.

This change ensures that when validation errors occur, the field names in the error messages correspond to the JSON tags of the fields, which is more user-friendly.

jan-zabloudil commented on supabase/migrations/20240328123715_projects.sql:
```diff
@@ -0,0 +1,27 @@
+                                     "name" text not null,
+                                     "description" text,
+                                     "release_alert_config" json,
+                                     "slack_config" json,
```
I understand your point, but the issue is that Supabase does not support transactions, which complicates scenarios where you want to insert data into multiple tables. 
To achieve this, I would need to write a PostgreSQL function, which is not very maintainable. In this case, I feel that having the configuration saved in JSON is sufficient, but I will discuss it further with Marek during our call. Thank you!

jan-zabloudil commented on supabase/migrations/20240328123715_projects.sql:
```diff
@@ -0,0 +1,27 @@
+
+grant truncate on table "public"."projects" to "service_role";
+
+grant update on table "public"."projects" to "service_role";
```
It was auto-generated migration. You are right, I fixed it

jan-zabloudil commented:
This branch is now rebased onto https://github.com/jan-zabloudil/release-manager/pull/16 because during review I also rebased it to work with more recent code.

jan-zabloudil commented on pkg/dberrors/dberrors.go:
```diff
@@ -46,10 +46,9 @@ func NewMultipleOrNoRecordsReturnedError() *DBError {
 
-func NewUnknownError(err error) *DBError {
+func NewUnknownError() *DBError {
 	return &DBError{
```
Not related to this PR but addresses an issue left over from a previous PR.

jan-zabloudil commented on service/model/user_role.go:
```diff
@@ -7,22 +7,25 @@ import (
-	UserRoleUser  UserRole = "user"
+	UserRoleAdmin   UserRole = "admin"
+	UserRoleManager UserRole = "manager"
+	UserRoleUser    UserRole = "user"
```
New role `Manager` introduces. Why?

- Admin  role - have access to organization settings (implemented in #14) and can create projects as well

- Manager role - does not have access to organization settings, but can create projects _(role for Engineering managers in STRV)_

Other naming option would be `SuperAdmin` and `Admin roles`.

If you approve the addition of this new role, I will also create an endpoint that enables admin users to assign the manager role to other users.

bolshakoff replied:
Interesting point! I'd definitely steer towards something more simple and obvious, for example:
- org admin
- project admin
- user
How about other similar platforms? What names are they using?

And do we expect to have more different items?

If yes, consider simply having users and rights (read/modify org/project).

jan-zabloudil commented on service/project.go:
```diff
@@ -0,0 +1,95 @@
+}
+
+func (s *ProjectService) Get(ctx context.Context, projectID uuid.UUID, authUserID uuid.UUID) (model.Project, error) {
+	// TODO add project member authorization
```
Authorization will be added in the next PR, where project members feature will be added.

jan-zabloudil commented on service/model/project.go:
```diff
@@ -0,0 +1,160 @@
+	RlsAlertCfgUpdate ReleaseAlertConfigUpdate
+}
+
+type ReleaseAlertConfig struct {
```
`ReleaseAlertConfig` defines the information that will be included in a release alerts for a given project. 

I chose 'ReleaseAlert' over 'ReleaseMessage' as I believe it’s a more suitable name

bolshakoff replied:
`Alert` sounds too negative. :)
Also agree that `message` is a bit too neutral and general.
How about `notification`?

jan-zabloudil replied:
Good point, let's go with `ReleaseNotification`

jan-zabloudil commented:
When discussing the release management app proposal, we considered two approaches for organizing releases:

**Project > Releases**
- originally proposed by @CermakM.
- each project represents a single application (e.g., Backend app for project X).
- one project = one GitHub repository.
- I suggest renaming it to `App > Releases`.
 
**Project > Apps > Releases**

I initially favored the second approach (Project > Apps > Releases) and built the entire proposal around this structure. But I have realized that it introduces unnecessary complexity without offering significant benefits. The first suggestion is much more flexible and aligns better with company workflows (including STRV).

**Why?**
Different teams typically handle different platforms for given project. For example, the Go team manages the backend solution, the iOS team handles the iOS app, and the Android team takes care of the Android app.

It makes sense for each Engineering Manager to create an app within ReleaseManager for which they are responsible. They can then invite engineers from their team with permissions to create releases. Managers can also invite engineers from other teams or PMs but only grant them read permissions. The manager will also connect the app to a GitHub repository and configure release notifications. 

But then, do we really need the project level at all? 
It seems that adding the project level would not bring any significant benefits and would only increase complexity. Adopting this approach would require managing permissions and settings on two levels. Or keep settings/permissions only on one level but significantly decrease the release manager app’s flexibility.

Since we are developing an open-source MVP, I would definitely choose the App > Releases solution.

CermakM commented on pkg/apierrors/apierrors.go:
```diff
@@ -54,7 +75,21 @@ func IsNotFoundError(err error) bool {
 	if errors.As(err, &apiErr) {
 		switch apiErr.Code {
-		case userNotFoundErrCode:
+		case userNotFoundErrCode, projectNotFoundErrCode:
```
Just a suggestion here, but as you can see, the `IsError` functions have all the same definition and the only thing they differ in are the error codes comparisons. Meaning you can simplify this flow by implementing one `IsErrorWithCode` function of the signiture `IsErrorWithCode(err error, code Code)`.

CermakM commented on pkg/responseerrors/responseerrors.go:
```diff
@@ -9,14 +9,16 @@ import (
-	methodNotAllowedErrCode       = "ERR_METHOD_NOT_ALLOWED"
-	invalidResourceIDErrCode      = "ERR_INVALID_RESOURCE_ID"
-	unknownErrCode                = "ERR_UNKNOWN"
+	notBearerTokenFormatErrorCode     = "ERR_TOKEN_NOT_BEARER_FORMAT"
```
I would rather structure the error codes as `errCodeX` rather than `xErrorCode`, this makes it easier to grep them both visually and via autocomplete.

CermakM commented on repository/model/project.go:
```diff
@@ -0,0 +1,93 @@
+	Name               string             `json:"name"`
+	SlackChannelID     string             `json:"slack_channel_id"`
+	ReleaseAlertConfig ReleaseAlertConfig `json:"release_alert_config"`
+	CreatedAt          iso8601.Time       `json:"created_at"`
```
Why not `time.Time`?

jan-zabloudil replied:
Thanks! Initially, I had problems with marshaling/unmarshaling into a time.Time variable. But that was probably due to some beginner's mistakes. It works fine with `time.Time` for `timestamp with timezone` column 👍

CermakM commented on repository/project.go:
```diff
@@ -0,0 +1,154 @@
+	}
+}
+
+func (r *ProjectRepository) Create(ctx context.Context, p svcmodel.Project) (svcmodel.Project, error) {
```
Is there any point in returning the same structure as passed as a parameter? To me it looks like this is an unnecessary overhead and simply returning an `error` would be enough.

CermakM commented on service/model/project.go:
```diff
@@ -0,0 +1,160 @@
+		Name:               name,
+		SlackChannelID:     slackChannelID,
+		ReleaseAlertConfig: cfg,
+		CreatedAt:          time.Now(),
```
Define a variable `now` and use it for both fields, otherwise they might not be equal, which is weird considering no update happened.

CermakM commented:
Thanks for the elaboration on the entity structure! I 💯 agree 👍 

Regarding the user permissions:

> AppAdmin / ProjectAdmin
>
> Can create apps
  Does not have access to organization settings
  This role would typically be assigned to Engineering Managers
  
In the current STRV setting, I believe it's okay for us to have only organization admin and project member structure. However, from a long term perspective, I would absolutely consider having a project admin. Especially if the organization-wide settings grows in complexity and might potentially hold sensitive info. Also, since project administrators may not be technically savvy, there is some risk in allowing them access to org-wide settings.

jan-zabloudil commented:
Thank you 👍 I'm reassigning the PR to myself and will implement the remaining functionalities for `apps`.

jan-zabloudil commented:
- Environments have been added.

To keep PRs smaller, the following features will be implemented in separate PRs:
- Project members
- Connection to GitHub repository (it would be good to merge PR #14  first to have access to the GitHub token).

In the end, I also decided to continue referring to it as `Project` instead of `App`. It is more general, but I think it’s not necessarily a bad thing. Moreover, `project members` works better than `app members`, etc.

jan-zabloudil commented on pkg/url/url.go:
```diff
@@ -0,0 +1,14 @@
+func IsAbsolute(raw string) bool {
+	u, err := url.Parse(raw)
+	if err != nil || u.Scheme == "" || u.Host == "" {
+		return false
```
I'm not sure about this... I at least wanted to perform a basic check to see if a URL is valid according to a 'human' definition of a "correct" URL (of course, many incorrect ones would still pass). 

Or it might be better to use some existing validator...

bolshakoff replied:
Yes, I think it can be done using the `validator` package.

jan-zabloudil replied:
Thanks, I am going to change it in separate PR once this one is merged 👍

jan-zabloudil commented:
Sorry, I've realized I need to implement a different approach to environments. Reassigning it to myself again and I'm on it.

jan-zabloudil commented:
> Sorry, I've realized I need to implement a different approach to environments. Reassigning it to myself again and I'm on it.

In STRV, all projects have development, staging, and production environments. However, I have decided not to have fixed environments to maintain flexibility, especially since I am developing an open source app.

1. When a manager creates a project, they can set up environments; the name field is required, while the URL and release fields are optional. _Potentially in STRV, we could automatically create development, staging, and production environments when a project is created_.

 2. When using the app, you can visit the environments page to see which releases are deployed to each environment, and you can quickly make changes.

![Projects _ Environments](https://github.com/jan-zabloudil/release-manager/assets/81581922/8d942fc1-5881-42e3-b6a2-796f4479a650)

You can also "deploy" a release to an environment from the release page.

![Projects _ Releases _ Add](https://github.com/jan-zabloudil/release-manager/assets/81581922/5fe340bd-13a6-434c-9b2e-f27d915756cc)

CermakM commented on service/model/environment.go:
```diff
@@ -0,0 +1,133 @@
+	RawURL    *string
+}
+
+func NewEnvironment(projectID uuid.UUID, name, rawURL string) (Environment, error) {
```
Why do I need to provide this URL? It should be accesible on the base url + id, shouldn't it?

jan-zabloudil replied:
Renamed to `ServiceURL` to improve clarity.

CermakM commented on service/environment.go:
```diff
@@ -0,0 +1,121 @@
+func (s *EnvironmentService) Create(ctx context.Context, ec model.EnvironmentCreation, authUserID uuid.UUID) (model.Environment, error) {
+	// TODO add project member authorization
+
+	_, err := s.projectSvc.Get(ctx, ec.ProjectID, authUserID)
```
Shouldn't this be only the admin's privilage?

CermakM commented on service/environment.go:
```diff
@@ -0,0 +1,121 @@
+func (s *EnvironmentService) Delete(ctx context.Context, projectID, envID uuid.UUID, authUserID uuid.UUID) error {
+	// TODO add project member authorization
+
+	_, err := s.projectSvc.Get(ctx, projectID, authUserID)
```
Same here, project admin only I would say.

CermakM commented on supabase/migrations/20240417150340_environments.sql:
```diff
@@ -0,0 +1,13 @@
+create table "public"."environments" (
+         "id" uuid not null default gen_random_uuid(),
+         "project_id" uuid not null references public.projects on delete cascade,
+         "name" text not null,
```
Should be unique per project.

CermakM commented on service/environment.go:
```diff
@@ -0,0 +1,121 @@
+	repository model.EnvironmentRepository
+}
+
+func NewEnvironmentService(authSvc model.AuthService, projectSvc model.ProjectService, repo model.EnvironmentRepository) *EnvironmentService {
```
I think it's an unnecessary overhead to have environment creation as a separate service. It's part of the project resource, doesn't have a seaparate resource endpoint, hence I would put it under the project service. Not that you don't have to have a service per resource either, rather structure it per domain.

jan-zabloudil commented on pkg/dberrors/dberrors.go:
```diff
@@ -6,10 +6,10 @@ import (
-	unknownErrCode             = "ERR_DB_UNKNOWN"
+	errCodeNotFound            = "ERR_DB_RESOURCE_NOT_FOUND"
+	errCodeCannotMapToSvcModel = "ERR_DB_CANNOT_MAP_TO_SVC_MODEL"
+	errCodeMultipleOrNoRecords = "ERR_DB_MULTIPLE_OR_NO_RECORDS_RETURNED"
```
In the end I dont need that error anymore, I removed it.

bolshakoff commented on service/project.go:
```diff
@@ -0,0 +1,233 @@
+	}
+
+	if err := s.validateEnvironmentNameUnique(ctx, env.ProjectID, env.Name); err != nil {
+		return model.Environment{}, err
```
Sorry, consider this resolved..

But actually..
I'd return the duplicate name error here; ideally simplify it to return `(*bool, error)`.
A bit more code, but more obvious, more readable.

bolshakoff replied:
I mean `s.validateEnvironmentNameUnique()`

jan-zabloudil replied:
Thank you for review, all very good points!

Just here I chose to return `(bool, error)`.
I understand that by returning `(*bool, error)`, you can distinguish the following cases based on bool value:
- true - the name is unique
- false - the name is not unique
- nil - it cannot be determined if the name is unique

But in case of `(bool, error)` I simply check if err is nil. If it isn’t, I know that the uniqueness couldn't be determined due to an error. If err is nil, I can use the bool value directly without needing to dereference it.

jan-zabloudil commented on service/model/environment.go:
```diff
@@ -119,7 +110,7 @@ func (e *Environment) Update(name, rawURL *string) error {
 
 	if rawURL != nil {
-		if err := e.SetURL(*rawURL); err != nil {
+		if err := e.SetServiceURL(*rawURL); err != nil {
```
> (optional) And actually, it might be even more clean to have the following signature:
> 
>     * `Update(u EnvironmentUpdate) (*Environment, error)`

Not sure if I get the point. What is the benefit of returning a pointer to the Environment in the case of a pointer receiver? Is it just for better readability?

```
env, err := env.Update(...)
```
